Update #r pacakges tooling RFC (#502)
fsharp/fslang-design: tooling/FST-1027-fsi-references.md

The following extensions to `#r` are proposed:

* `#r "[dependency manager]: <dependency manager command>"` like
   * `#r "paket: <paket command>"`
   * `#r "nuget: <package-name>, <package-version>"`

This extends the "language" of `#r` to support NuGet packages, Paket dependencies, and potentially more.

# Motivation
[motivation]: #motivation

There is a [strong desire](https://github.com/fsharp/fslang-suggestions/issues/542) to reference packages via `#r` instead of assemblies.  There has also been [experimental work](https://github.com/Microsoft/visualfsharp/pull/2483) to support Paket in this way.

# Detailed design
[design]: #detailed-design

To begin, existing behavior with `#r` will remain unchanged for FSI.  That is, if you reference an assembly today via `#r`, it will always continue to work in the same way that it always has.

> .NET Core necessitates the introduction of `#r "nuget: name, version"` and `#r "paket: paket-command"` to simplify referencing dependencies in F# script files for .NET Core.  The intention is that FSI Scripts in the future will always specify dependencies via `#r "nuget` or `#r "paket` (or a future extended command, like `#r "project"`), allowing only FSI to be concerned with referencing the correct assemblies.

Reviewer: ```suggestion
.NET Core necessitates the introduction of `#r "nuget: name, version"` to simplify referencing dependencies in F# script files for .NET Core.  The intention is that FSI Scripts in the future will always specify dependencies via `#r "nuget:...` (or other extensions, like `#r "paket:..."`, `#r "frameworkreference:...", `#r "project:..."`, etc.), allowing only FSI to be concerned with referencing the correct assemblies.
```

Paket is not built in and would have to be an extension. Also added some others that have come up as a request

As above, if they somehow specify something which is already loaded, we should just let FSI attempt to load it and generate the error.

### Referencing packages containing native DLLs 

Referencing packages containing native DLLs is supported with some limitations.  

The support is achieved by  adding an event handler to the .NET Core event `AssemblyLoadContext.Default.ResolvingUnmanagedDll`, which is triggered when resolving an unmanaged assembly in the context of a .NET assembly (e.g. a DllImport).   

> This handler consults current architecture and platform settings plus resolved package metadata and files across all dynamically referenced packages to look for a matching native DLL and then dynamically loads that DLL using an internal NativeAssemblyLoadContext that implements `LoadNativeLibrary` via `LoadUnmanagedDllFromPath`.  

Reviewer: ```suggestion
This handler consults current architecture and platform settings plus resolved package metadata and files across all dynamically referenced packages to look for a matching native DLL and then dynamically loads that DLL using an internal `NativeAssemblyLoadContext` that implements `LoadNativeLibrary` via `LoadUnmanagedDllFromPath`.  
```

If a handler key (such as `nuget` or `paket`) is found several times, report a warning showing location of assemblies and showing which one was picked (we apply same order of precedence as for finding the assemblies).

# Tooling

> Language service and FCS tooling must be efficient given incremental editing of files.

Reviewer: I suppose more detail needs to be given here, perhaps clarifying that design-time resolution of scripts is orthogonal to runtime and carries some concerns:

* Editor performance
* A delay in various tooling due to nuget resolution happening in the background
* Ensuring that we don't re-resolve things unnecessarily
* Future extensions to indicate background work is happening, like how project integration tooling does today

Author: ok
